### PR TITLE
webtransport: move certhash verification to the client

### PR DESCRIPTION
**Pull request by marten-seemann**
As suggested in https://github.com/libp2p/specs/pull/455.

This PR builds on top #1750. **Please only review the last commit**. I'll rebase this PR once #1750 is merged.

### CODE REVIEW

File: p2p/transport/webtransport/transport.go [hunk 1]
```diff
@@ -196,32 +198,62 @@ func (t *transport) upgrade(ctx context.Context, sess *webtransport.Session, p p
 
 	// Now run a Noise handshake (using early data) and send all the certificate hashes that we would have accepted.
 	// The server will verify that it advertised all of these certificate hashes.
-	msg := pb.WebTransport{CertHashes: make([][]byte, 0, len(certHashes))}
-	for _, certHash := range certHashes {
-		h, err := multihash.Encode(certHash.Digest, certHash.Code)
+	var verified bool
+	n, err := t.noise.WithSessionOptions(noise.EarlyData(newEarlyDataReceiver(func(b []byte) error {
+		ch, err := decodeCertHashesFromProtobuf(b)
```
Comment (MarcoPolo) [inline p2p/transport/webtransport/transport.go:203]: Nit: maybe something more descriptive than `ch`?

File: p2p/transport/webtransport/transport.go [hunk 2]
```diff
@@ -196,32 +198,62 @@ func (t *transport) upgrade(ctx context.Context, sess *webtransport.Session, p p
...
+	}
 	return &connSecurityMultiaddrs{
 		ConnSecurity:   c,
 		ConnMultiaddrs: &connMultiaddrs{local: local, remote: remote},
 	}, nil
 }
 
+func decodeCertHashesFromProtobuf(b []byte) ([]multihash.DecodedMultihash, error) {
+	var msg pb.WebTransport
+	if err := msg.Unmarshal(b); err != nil {
```
Comment (MarcoPolo) [inline p2p/transport/webtransport/transport.go:242]: should we error if len(b) == 0? (The unmarshal will not fail this I think)

Comment (MarcoPolo) [reply p2p/transport/webtransport/transport.go:242]: might be okay. A server sending an empty early data can be interpreted to mean that there is no certhash. Just a bit surprising

**Review by MarcoPolo — APPROVED** [review]

File: p2p/transport/webtransport/transport.go (hunk 1 above)
Comment (marten-seemann) [reply p2p/transport/webtransport/transport.go]: `decodedCertHashes`.

**Pull request merged**